File: tests/fakevim.py
import re


class FakeError(Exception):
    pass


# WHY: vim rejects counts that are not a number with an optional s, m, h, d or f unit.
COUNT_PATTERN = re.compile(r"^(\d+)[smhdf]?$")


def as_numbers(value):
    if isinstance(value, dict):
        return {key: as_numbers(item) for key, item in value.items()}

    if isinstance(value, list):
        return [as_numbers(item) for item in value]

    if isinstance(value, str) and value.lstrip("-").isdigit():
        return int(value)

    return value


class FakeBuffer:
    def __init__(self, number, lines=None, name=""):
        self.number = number
        self.name = name
        self.lines = list(lines if lines is not None else [])
        self.options = {}

    def __getitem__(self, index):
        return self.lines[index]

    def __setitem__(self, index, value):
        if isinstance(index, slice):
            self.lines[index] = list(value)
        else:
            self.lines[index] = value

    def __len__(self):
        return len(self.lines)

    def __repr__(self):
        return f"<FakeBuffer {self.number} {self.name!r}>"


class FakeWindow:
    def __init__(self, buffer):
        self.buffer = buffer

    def __repr__(self):
        return f"<FakeWindow buffer={self.buffer.number}>"


class FakeCurrent:
    def __init__(self, vim):
        self._vim = vim

    @property
    def window(self):
        return self._vim._current_window

    @window.setter
    def window(self, window):
        self._vim._current_window = window

    @property
    def buffer(self):
        return self._vim._current_window.buffer

    @buffer.setter
    def buffer(self, buffer):
        self._vim._current_window.buffer = buffer


class UndoHistory:
    def __init__(self, states, times=None):
        self.states = [list(lines) for lines in states]
        self.times = times or [str(1627784659 + i * 60) for i in range(len(states))]
        self.seq = len(states) - 1

    @property
    def lines(self):
        return self.states[self.seq]

    def undo(self, seq):
        self.seq = max(0, min(int(seq), len(self.states) - 1))

    def earlier(self, count=1):
        self.undo(self.seq - int(count))

    def later(self, count=1):
        self.undo(self.seq + int(count))

    def undotree(self):
        return {
            "seq_last": str(len(self.states) - 1),
            "seq_cur": str(self.seq),
            "entries": [{"seq": str(i), "time": self.times[i]} for i in range(1, len(self.states))],
        }


class FakeVim:
    error = FakeError

    def __init__(
        self, history, filetype="python", name="source.py", numeric_eval=False, winbar=True
    ):
        self.history = history
        self.numeric_eval = numeric_eval
        self.winbar = winbar
        self.filetype = filetype
        self.vars = {}
        self.commands = []
        self.messages = []
        self._next_bufnr = 1
        source = self._new_buffer_object(list(history.lines), name)
        self.buffers = [source]
        self.source_buffer = source
        self.windows = [FakeWindow(source)]
        self._current_window = self.windows[0]
        self.current = FakeCurrent(self)

    def close_window(self, window):
        self.windows.remove(window)
        if self._current_window is window:
            self._current_window = self.windows[0]

        if window.buffer.options.get("bufhidden") == "wipe":
            self.buffers.remove(window.buffer)

    def window_of_buffer(self, number):
        return next(w for w in self.windows if w.buffer.number == int(number))

    def _new_buffer_object(self, lines=None, name=""):
        buffer = FakeBuffer(self._next_bufnr, lines, name)
        self._next_bufnr += 1
        return buffer

    def buffer_named(self, number):
        return next(b for b in self.buffers if b.number == int(number))

    @property
    def diff_buffer(self):
        return self.buffer_named(self.vars["t:diffundo_diff_bn"])

    def _sync_source_buffer(self):
        self.source_buffer[:] = list(self.history.lines)

    def eval(self, expression):
        value = self._eval(expression)
        # WHY: neovim's python host returns numbers where vim's returns strings.
        return as_numbers(value) if self.numeric_eval else value

    def _eval(self, expression):
        if expression == "changenr()":
            return str(self.history.seq)

        if expression == "undotree()":
            return self.history.undotree()

        if expression == "&filetype":
            return self.filetype

        if expression == "exists('+winbar')":
            return "1" if self.winbar else "0"

        if expression.startswith("t:"):
            if expression not in self.vars:
                raise FakeError(f"E121: Undefined variable: {expression}")
            return self.vars[expression]

        if expression.startswith("bufexists("):
            inner = expression[len("bufexists(") : -1]
            number = self._eval(inner) if inner.startswith("t:") else inner
            return "1" if any(b.number == int(number) for b in self.buffers) else "0"

        raise FakeError(f"unsupported eval: {expression}")

    def command(self, command):
        self.commands.append(command)
        stripped = command
        while stripped.startswith("silent "):
            stripped = stripped[len("silent ") :]

        if stripped.startswith("/") or stripped == "diffupdate":
            return

        head, _, rest = stripped.partition(" ")
        handler = self._command_handlers().get(head)
        if handler is None:
            raise FakeError(f"unsupported command: {command}")
        handler(head, rest.strip())

    def _command_handlers(self):
        return {
            "let": self._command_let,
            "undo": self._command_undo,
            "earlier": self._command_earlier_later,
            "later": self._command_earlier_later,
            "file": self._command_file,
            "enew": self._command_enew,
            "vert": self._command_vert,
            "setlocal": self._command_setlocal,
        }

    def _command_let(self, head, rest):
        name, _, value = rest.partition("=")
        name, value = name.strip(), value.strip()
        if name.startswith("&l:"):
            self.current.buffer.options[name[len("&l:") :]] = value.strip("'")
            return

        self.vars[name] = value

    def _command_undo(self, head, rest):
        self.history.undo(rest)
        self._sync_source_buffer()

    def _command_earlier_later(self, head, rest):
        match = COUNT_PATTERN.match(rest or "1")
        if match is None:
            raise FakeError(f"Vim({head}):E475: Invalid argument: {rest}")

        getattr(self.history, head)(match.group(1))
        self._sync_source_buffer()

    def _command_file(self, head, rest):
        self.current.buffer.name = rest

    def _command_enew(self, head, rest):
        buffer = self._new_buffer_object()
        self.buffers.append(buffer)
        self.current.window.buffer = buffer

    def _command_vert(self, head, rest):
        if rest != "diffsplit":
            raise FakeError(f"unsupported command: vert {rest}")
        window = FakeWindow(self.current.buffer)
        self.windows.insert(0, window)
        self._current_window = window

    def _command_setlocal(self, head, rest):
        name, _, value = rest.partition("=")
        self.current.buffer.options[name] = value or True

File: spec/fakevim.lua
local M = {}

---@param lines string[]
---@return string[]
local function copy(lines)
  local result = {}
  for index, line in ipairs(lines) do
    result[index] = line
  end
  return result
end

local History = {}
History.__index = History

---@param states string[][]
---@param times integer[]|nil
function M.history(states, times)
  local self = setmetatable({}, History)
  self.entries = {}
  for index, lines in ipairs(states) do
    local seq = index - 1
    self.entries[seq] = {
      lines = copy(lines),
      parent = math.max(0, seq - 1),
      time = times and times[index] or 1627784659 + seq * 60,
    }
  end
  self.seq_last = #states - 1
  self.seq = self.seq_last
  return self
end

---@param parent integer
---@param lines string[]
---@param opts { time?: integer, save?: integer }|nil
---@return integer
function History:branch(parent, lines, opts)
  local options = opts or {}
  local seq = self.seq_last + 1
  self.entries[seq] = {
    lines = copy(lines),
    parent = parent,
    time = options.time or (1627784659 + seq * 60),
    save = options.save,
  }
  self.seq_last = seq
  self.seq = seq
  return seq
end

---@return string[]
function History:lines()
  return self.entries[self.seq].lines
end

---@param seq integer
function History:undo(seq)
  self.seq = math.max(0, math.min(seq, self.seq_last))
end

---@param count integer
function History:earlier(count)
  self:undo(self.seq - count)
end

---@param count integer
function History:later(count)
  self:undo(self.seq + count)
end

---@param self table
---@param parent integer
---@return integer[]
local function children(self, parent)
  local result = {}
  for seq = self.seq_last, 1, -1 do
    if self.entries[seq].parent == parent then
      table.insert(result, seq)
    end
  end
  return result
end

---@param list integer[]
---@return integer[]
local function tail(list)
  local result = {}
  for index = 2, #list do
    result[index - 1] = list[index]
  end
  return result
end

---@param self table
---@param seq integer
---@param siblings integer[]
---@return table[]
local function chain(self, seq, siblings)
  local entry = self.entries[seq]
  local head = { seq = seq, time = entry.time, save = entry.save }
  if siblings[1] then
    head.alt = chain(self, siblings[1], tail(siblings))
  end
  local list = { head }
  local kids = children(self, seq)
  if kids[1] then
    for _, item in ipairs(chain(self, kids[1], tail(kids))) do
      table.insert(list, item)
    end
  end
  return list
end

function History:undotree()
  local kids = children(self, 0)
  local entries = kids[1] and chain(self, kids[1], tail(kids)) or {}
  return { seq_last = self.seq_last, seq_cur = self.seq, entries = entries }
end

local Fake = {}
Fake.__index = Fake

---@param self table
---@return table
local function current_buffer(self)
  return self.buffers[self.windows[self.current_win].buf]
end

---@param self table
---@param bufnr integer|nil
---@return integer|nil
local function window_of_buffer(self, bufnr)
  for _, win in ipairs(self.win_order) do
    if self.windows[win].buf == bufnr then
      return win
    end
  end
  return nil
end

---@param self table
---@param lines string[]|nil
---@param name string|nil
---@return integer
local function new_buffer(self, lines, name)
  local bufnr = self.next_bufnr
  self.next_bufnr = bufnr + 1
  self.buffers[bufnr] =
    { number = bufnr, lines = copy(lines or {}), name = name or "", options = {} }
  return bufnr
end

---@param self table
local function sync_source(self)
  self.buffers[self.source_bn].lines = copy(self.history:lines())
end

---@param self table
---@return table
local function commands(self)
  return {
    undo = function(rest)
      self.history:undo(tonumber(rest))
      sync_source(self)
    end,
    earlier = function(rest)
      local count = (rest == "" and "1" or rest):match("^(%d+)[smhdf]?$")
      if count == nil then
        error("Vim(earlier):E475: Invalid argument: " .. rest, 0)
      end
      self.history:earlier(tonumber(count))
      sync_source(self)
    end,
    later = function(rest)
      local count = (rest == "" and "1" or rest):match("^(%d+)[smhdf]?$")
      if count == nil then
        error("Vim(later):E475: Invalid argument: " .. rest, 0)
      end
      self.history:later(tonumber(count))
      sync_source(self)
    end,
    diffupdate = function() end,
    enew = function()
      self.windows[self.current_win].buf = new_buffer(self)
    end,
    vert = function(rest)
      if rest ~= "diffsplit" then
        error("unsupported command: vert " .. rest, 0)
      end
      local win = self.next_win
      self.next_win = win + 1
      self.windows[win] =
        { buf = self.windows[self.current_win].buf, options = {}, cursor = { 1, 0 } }
      table.insert(self.win_order, 1, win)
      self.current_win = win
    end,
  }
end

---@param self table
---@param command string
local function run_command(self, command)
  table.insert(self.commands, command)
  local stripped = command:gsub("^silent ", "")
  local head, rest = stripped:match("^(%S+)%s*(.*)$")
  local handler = commands(self)[head]
  if handler == nil then
    error("unsupported command: " .. command, 0)
  end
  handler(rest)
end

---@param self table
---@param win integer
---@return table
local function window(self, win)
  local resolved = win == 0 and self.current_win or win
  local found = self.windows[resolved]
  if found == nil then
    error("Invalid window id: " .. tostring(win), 0)
  end
  return found
end

---@param self table
---@return table
local function api(self)
  return {
    nvim_get_current_win = function()
      return self.current_win
    end,
    nvim_win_is_valid = function(win)
      return self.windows[win] ~= nil
    end,
    nvim_win_get_cursor = function(win)
      return copy(window(self, win).cursor)
    end,
    nvim_win_set_cursor = function(win, pos)
      window(self, win).cursor = { pos[1], pos[2] }
    end,
    nvim_tabpage_list_wins = function()
      return copy(self.win_order)
    end,
    nvim_win_get_buf = function(win)
      return self.windows[win].buf
    end,
    nvim_set_current_win = function(win)
      self.current_win = win
    end,
    nvim_get_current_buf = function()
      return current_buffer(self).number
    end,
    nvim_buf_is_valid = function(bufnr)
      return self.buffers[bufnr] ~= nil
    end,
    nvim_buf_get_lines = function()
      return copy(current_buffer(self).lines)
    end,
    nvim_buf_set_lines = function(_, _, _, _, lines)
      current_buffer(self).lines = copy(lines)
    end,
    nvim_buf_set_name = function(_, name)
      current_buffer(self).name = name
    end,
  }
end

---@param self table
---@return table
local function fn(self)
  return {
    changenr = function()
      return self.history.seq
    end,
    undotree = function()
      return self.history:undotree()
    end,
    escape = function(text)
      return text
    end,
    search = function(pattern)
      table.insert(self.searches, pattern)
      return 1
    end,
  }
end

---@param get fun(): table
---@return table
local function option_proxy(get)
  return setmetatable({}, {
    __index = function(_, name)
      return get()[name]
    end,
    __newindex = function(_, name, value)
      get()[name] = value
    end,
  })
end

---@param history table
---@param opts { filetype?: string, name?: string }|nil
---@return table
function M.new(history, opts)
  local options = opts or {}
  local self = setmetatable({}, Fake)
  self.history = history
  self.buffers = {}
  self.windows = {}
  self.win_order = {}
  self.next_bufnr = 1
  self.next_win = 1000
  self.notifications = {}
  self.commands = {}
  self.searches = {}
  self.t = {}
  self.log = { levels = { ERROR = 4, INFO = 2 } }

  self.source_bn = new_buffer(self, history:lines(), options.name or "source.lua")
  self.buffers[self.source_bn].options.filetype = options.filetype or "lua"
  local win = self.next_win
  self.next_win = win + 1
  self.windows[win] = { buf = self.source_bn, options = {}, cursor = { 1, 0 } }
  self.win_order = { win }
  self.current_win = win

  self.api = api(self)
  self.fn = fn(self)
  self.bo = option_proxy(function()
    return current_buffer(self).options
  end)
  self.wo = option_proxy(function()
    return self.windows[self.current_win].options
  end)
  self.cmd = function(command)
    run_command(self, command)
  end
  self.notify = function(message)
    table.insert(self.notifications, message)
  end
  self.keycode = function(keys)
    return keys
  end
  self.regex = function(pattern)
    local literal = pattern:gsub("^\\V", "")
    return {
      match_str = function(_, line)
        local start, stop = line:find(literal, 1, true)
        if start == nil then
          return nil
        end
        return start - 1, stop
      end,
    }
  end
  return self
end

---@return table
function Fake:source_buffer()
  return self.buffers[self.source_bn]
end

---@return table
function Fake:diff_buffer()
  return self.buffers[self.t.diffundo_diff_bn]
end

---@return table
function Fake:diff_window()
  return self.windows[window_of_buffer(self, self.t.diffundo_diff_bn)]
end

---@return table
function Fake:current_buffer()
  return current_buffer(self)
end

---@param bufnr integer
---@return integer|nil
function Fake:window_of_buffer(bufnr)
  return window_of_buffer(self, bufnr)
end

---@param win integer
function Fake:close_window(win)
  for index, candidate in ipairs(self.win_order) do
    if candidate == win then
      table.remove(self.win_order, index)
    end
  end
  local closed = self.windows[win]
  self.windows[win] = nil
  if self.current_win == win then
    self.current_win = self.win_order[1]
  end
  if self.buffers[closed.buf].options.bufhidden == "wipe" then
    self.buffers[closed.buf] = nil
  end
end

-- selene: allow(global_usage)
function Fake:install()
  _G.vim = self
end

---@return string
function Fake:last_notification()
  return self.notifications[#self.notifications] or ""
end

return M
